src/clp_ffi_js/ir/StructuredIrUnitHandler.cpp

```diff
@@ -0,0 +1,160 @@
+
+auto StructuredIrUnitHandler::handle_log_event(StructuredLogEvent&& log_event
+) -> clp::ffi::ir_stream::IRErrorCode {
+    auto const& id_value_pairs{log_event.get_node_id_value_pairs()};
+    auto const timestamp = get_timestamp(id_value_pairs);
+    auto const log_level = get_log_level(id_value_pairs);
+
+    m_deserialized_log_events->emplace_back(std::move(log_event), log_level, timestamp);
+
+    return clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+}
+
+auto StructuredIrUnitHandler::handle_utc_offset_change(
+        [[maybe_unused]] clp::UtcOffset utc_offset_old,
+        [[maybe_unused]] clp::UtcOffset utc_offset_new
+) -> clp::ffi::ir_stream::IRErrorCode {
+    SPDLOG_WARN("UTC offset change packets aren't handled currently.");
+    return clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+}
+
+auto StructuredIrUnitHandler::handle_schema_tree_node_insertion(
+        clp::ffi::SchemaTree::NodeLocator schema_tree_node_locator
+) -> clp::ffi::ir_stream::IRErrorCode {
+    ++m_current_node_id;
+
+    auto const& key_name{schema_tree_node_locator.get_key_name()};
+    if (key_name == m_log_level_key) {
+        m_log_level_node_id.emplace(m_current_node_id);
+    } else if (key_name == m_timestamp_key) {
+        m_timestamp_node_id.emplace(m_current_node_id);
+    }
+
+    return clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+}
+
+auto StructuredIrUnitHandler::handle_end_of_stream() -> clp::ffi::ir_stream::IRErrorCode {
+    return clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+}
+
+auto StructuredIrUnitHandler::get_log_level(
+        StructuredLogEvent::NodeIdValuePairs const& id_value_pairs
+) const -> LogLevel {
+    LogLevel log_level{LogLevel::NONE};
+
+    if (false == m_log_level_node_id.has_value()) {
+        return log_level;
+    }
+    auto const& optional_log_level_value{id_value_pairs.at(m_log_level_node_id.value())};
+    if (false == optional_log_level_value.has_value()) {
+        return log_level;
+    }
+    auto const log_level_value = optional_log_level_value.value();
+
+    if (log_level_value.is<std::string>()) {
+        auto const& log_level_str = log_level_value.get_immutable_view<std::string>();
+        log_level = parse_log_level(log_level_str);
+    } else if (log_level_value.is<clp::ffi::value_int_t>()) {
+        auto const& log_level_int = log_level_value.get_immutable_view<clp::ffi::value_int_t>();
+        if (log_level_int >= clp::enum_to_underlying_type(cValidLogLevelsBeginIdx)
+            && log_level_int < clp::enum_to_underlying_type(LogLevel::LENGTH))
+        {
+            log_level = static_cast<LogLevel>(log_level_int);
+        }
+    } else {
+        auto log_event_idx = m_deserialized_log_events->size();
+        SPDLOG_ERROR(
+                "Authoritative log level's value is not an int or string for log event index {}",
+                log_event_idx
+        );
+    }
+
+    return log_level;
+}
+
+auto StructuredIrUnitHandler::get_timestamp(
```

Comment:
This may be deduplicated with `StructuredIrUnitHandler::get_log_level`, but for the sake of simplicity let's leave this for now.